zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java

```diff
@@ -1032,9 +1032,9 @@ public Vote lookForLeader() throws InterruptedException {
                          * together.
                          */
                         if (n.electionEpoch == logicalclock.get()) {
-                            recvset.put(n.sid, new Vote(n.leader, n.zxid, n.electionEpoch, n.peerEpoch));
+                            recvset.put(n.sid, new Vote(n.leader, n.zxid, n.electionEpoch, n.peerEpoch, n.state));
                             voteSet = getVoteTracker(recvset, new Vote(n.version, n.leader, n.zxid, n.electionEpoch, n.peerEpoch, n.state));
-                            if (voteSet.hasAllQuorums() && checkLeader(outofelection, n.leader, n.electionEpoch)) {
+                            if (voteSet.hasAllQuorums() && checkLeader(recvset, n.leader, n.electionEpoch)) {
```

Comment:
it might worth to add some comments at the declaration point of both recvset and outofelection, stating their intended uses. this will help clarify why this change was made in this pull request, also help future readers of the code understand the logic better.

Reply:
Thanks for your input @hanm. I introduced some comments at the declaration point of both `recvset` and `outofelection`, as well as, at the point where votes are being added in `outofelection`. What do you think?

Reply:
@insumity this seems to be a reasonable change to me, one question is why we add the n.state at line 1047, what's that used for?

Reply:
@lvfangmin in the [general case](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java#L999) (i.e., received notification is in a `LOOKING` state), `n.state` is not needed. 

However, note that we add `n.state` in the case that `FOLLOWING` or `LEADING` notifications are [received](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java#L1029). Assume that in such a scenario a notification from the leader is received and hence `n.state == LEADING`. If we do not add `n.state` in this case, `n.state` will be the default value of `LOOKING`. The problem with this is that when [`checkLeader`](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java#L1037) is called, and the state of the leader vote is subsequently [checked](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java#L793) it will not be in a `LEADING` state but it will be in a `LOOKING` state. Therefore, the `checkLeader` test will fail.

This was not a problem before because `checkLeader` was called in the `outofelection` set and the votes added in `outofelection` contained `n.state` as can be seen [here](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java#L1049).

Reply:
Make sense to me. @lvfangmin what do you think?

Reply:
Thanks for the details, I missed that when reviewing last time, this makes sense to me.